File: py/aoc08.py
from collections import namedtuple
import sys

import aoclib

Node = namedtuple('Node', ['metadata', 'children'])


def parse_node(stream):
    child_count = int(next(stream))
    metadata_count = int(next(stream))
    children = [parse_node(stream) for i in range(0, child_count)]
    metadata = [int(next(stream)) for i in range(0, metadata_count)]
    return Node(metadata, children)


def sum_metadata(node):
    child_sums = [sum_metadata(node) for node in node.children]
    return sum(node.metadata) + sum(child_sums)


def parse_stream(f):
    return f.read().split()


def main():
    stream = aoclib.int_stream(sys.stdin)
    answer = sum_metadata(parse_node(stream))
    print(answer)


if __name__ == '__main__':
    main()

File: ts/aoc08.ts
import * as _ from 'lodash';
import { readStdin } from './aoclib';

type Node = {
    metadata: Array<number>,
    children: Array<Node>,
};

const parseNode = (stream: Iterator<number>): Node => {
    const childCount = stream.next().value;
    const metadataCount = stream.next().value;
    const children = _.times(childCount, () => parseNode(stream));
    const metadata = _.times(metadataCount, () => stream.next().value);
    return { metadata, children };
};

const sumMetadata = (node: Node): number => {
    const childSums = _.map(node.children, sumMetadata);
    return _.sum(node.metadata) + _.sum(childSums);
};

const main = () => {
    const input = readStdin();
    const stream = _.map(input.split(' '), v => Number(v));
    const node = parseNode(stream[Symbol.iterator]());
    const answer = sumMetadata(node);
    console.log(answer);
};

main();
